model-optimizer/extensions/back/MatMulNormalizer.py

```diff
@@ -142,3 +145,75 @@ def replace_pattern(graph: Graph, match: dict):
             src = port.get_source()
             port.get_connection().set_source(transpose_copy.out_port(0))
             src.connect(start_port)
+
+
+class SmartReshape_HC_Reshape_MatMul(BackReplacementPattern):
+    """
+    Relaxes hard-coded input of Reshape in such sub-graphs:
+
+    input_1     Constant
+        \       /
+        Reshape    input_2
+           \       /
+              MatMul
+                |
+    """
+    enabled = True
+    force_clean_up = True
+
+    def run_after(self):
+        return [MatMulConstTransposesExtraction]
+
+    def pattern(self):
+        return dict(
+            nodes=[
+                ('output_shape', dict(type='Const')),
+                ('output_shape_d', dict()),
+                ('reshape', dict(type='Reshape')),
+                ('reshape_d', dict()),
+                ('other_input', dict(type=lambda t: t not in ['Reshape', 'Transpose'])),
+                ('other_input_d', dict()),
+                ('matmul', dict(type='MatMul')),
+            ],
+            edges=[
+                ('output_shape', 'output_shape_d'),
+                ('output_shape_d', 'reshape', {'in': 1}),
+                ('reshape', 'reshape_d'),
+                ('reshape_d', 'matmul'),
+                ('other_input', 'other_input_d'),
+                ('other_input_d', 'matmul'),
+            ]
+        )
+
+    def replace_pattern(self, graph: Graph, match: dict):
+        matmul = match['matmul']
+        reshape = match['reshape']
+        other_input_port_idx = 0 if match['matmul'].in_port(0).get_source().node.id == match['other_input'].id else 1
+        shape_source = match['matmul'].in_port(other_input_port_idx).get_source()
+        initial_reshape_pattern = reshape.in_port(1).data.get_value()
+        if len(initial_reshape_pattern) != 2:
+            return
+
+        reshape_is_A_input = matmul.in_port(0).get_source().node.id == reshape.id
+        if reshape_is_A_input:
+            idx = -1 if matmul.transpose_b else -2
```

Comment:
Why do we need negative values and `get_canonical_axis_index` if the transformation work for 2D tensors only?